File: src/main/java/com/yupi/airouter/controller/ImageController.java
package com.yupi.airouter.controller;

import cn.hutool.core.util.StrUtil;
import com.mybatisflex.core.paginate.Page;
import com.yupi.airouter.annotation.AuthCheck;
import com.yupi.airouter.common.BaseResponse;
import com.yupi.airouter.common.ResultUtils;
import com.yupi.airouter.constant.UserConstant;
import com.yupi.airouter.exception.BusinessException;
import com.yupi.airouter.exception.ErrorCode;
import com.yupi.airouter.model.dto.image.ImageGenerationRequest;
import com.yupi.airouter.model.dto.image.ImageGenerationResponse;
import com.yupi.airouter.model.entity.ApiKey;
import com.yupi.airouter.model.entity.ImageGenerationRecord;
import com.yupi.airouter.model.entity.User;
import com.yupi.airouter.service.ApiKeyService;
import com.yupi.airouter.service.ImageGenerationService;
import com.yupi.airouter.service.UserService;
import io.swagger.v3.oas.annotations.Operation;
import jakarta.annotation.Resource;
import jakarta.servlet.http.HttpServletRequest;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.*;

/**
 * 图片生成控制器
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
@RestController
@RequestMapping("/v1/images")
@Slf4j
public class ImageController {

    @Resource
    private ImageGenerationService imageGenerationService;

    @Resource
    private UserService userService;

    @Resource
    private ApiKeyService apiKeyService;

    /**
     * 生成图片（OpenAI 兼容接口）
     */
    @PostMapping("/generations")
    @Operation(summary = "生成图片")
    public ImageGenerationResponse generateImage(
            @RequestBody ImageGenerationRequest request,
            @RequestHeader(value = "Authorization", required = false) String authorization,
            HttpServletRequest httpRequest) {

        Long userId = null;
        Long apiKeyId = null;

        // 认证：优先使用 API Key，否则使用 Session
        if (StrUtil.isNotBlank(authorization) && authorization.startsWith("Bearer ")) {
            String apiKeyValue = authorization.substring(7);
            ApiKey apiKey = apiKeyService.getByKeyValue(apiKeyValue);

            if (apiKey == null) {
                throw new BusinessException(ErrorCode.NO_AUTH_ERROR, "API Key 无效或已失效");
            }
            if (!"active".equals(apiKey.getStatus())) {
                throw new BusinessException(ErrorCode.NO_AUTH_ERROR, "API Key 已被禁用");
            }

            userId = apiKey.getUserId();
            apiKeyId = apiKey.getId();
        } else {
            // 尝试从 Session 获取用户信息
            try {
                User loginUser = userService.getLoginUser(httpRequest);
                userId = loginUser.getId();
            } catch (BusinessException e) {
                throw new BusinessException(ErrorCode.NOT_LOGIN_ERROR, "请先登录或提供有效的 API Key");
            }
        }

        String clientIp = httpRequest.getRemoteAddr();

        // 调用服务生成图片
        return imageGenerationService.generateImage(request, userId, apiKeyId, clientIp);
    }

    /**
     * 获取我的图片生成记录
     */
    @GetMapping("/my/records")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "获取我的图片生成记录")
    public BaseResponse<Page<ImageGenerationRecord>> getMyRecords(
            @RequestParam(defaultValue = "1") int pageNum,
            @RequestParam(defaultValue = "10") int pageSize,
            HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);
        Page<ImageGenerationRecord> page = imageGenerationService.listUserRecords(
                loginUser.getId(), pageNum, pageSize);
        return ResultUtils.success(page);
    }
}

File: go-backend/internal/controller/image_controller.go
package controller

import (
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/yupi/airouter/go-backend/internal/common"
	"github.com/yupi/airouter/go-backend/internal/errno"
	"github.com/yupi/airouter/go-backend/internal/model/dto"
	"github.com/yupi/airouter/go-backend/internal/service"
)

type ImageController struct {
	imageService  *service.ImageGenerationService
	userService   *service.UserService
	apiKeyService *service.ApiKeyService
}

func NewImageController(
	imageService *service.ImageGenerationService,
	userService *service.UserService,
	apiKeyService *service.ApiKeyService,
) *ImageController {
	return &ImageController{
		imageService:  imageService,
		userService:   userService,
		apiKeyService: apiKeyService,
	}
}

func (i *ImageController) GenerateImage(c *gin.Context) {
	var request dto.ImageGenerationRequest
	if err := c.ShouldBindJSON(&request); err != nil {
		common.Error(c, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}

	var userID *int64
	var apiKeyID *int64
	authorization := strings.TrimSpace(c.GetHeader("Authorization"))
	if strings.HasPrefix(authorization, "Bearer ") {
		apiKeyValue := strings.TrimSpace(strings.TrimPrefix(authorization, "Bearer "))
		apiKey, err := i.apiKeyService.GetByKeyValue(apiKeyValue)
		if err != nil || apiKey == nil {
			common.Error(c, errno.NoAuthError.Code, "API Key 无效或已失效")
			return
		}
		if apiKey.Status != "active" {
			common.Error(c, errno.NoAuthError.Code, "API Key 已被禁用")
			return
		}
		userID = &apiKey.UserID
		apiKeyID = &apiKey.ID
	} else {
		loginUser, err := i.userService.GetLoginUser(c)
		if err != nil {
			common.Error(c, errno.NotLoginError.Code, "请先登录或提供有效的 API Key")
			return
		}
		userID = &loginUser.ID
	}

	response, err := i.imageService.GenerateImage(request, userID, apiKeyID, c.ClientIP())
	if err != nil {
		if bizErr, ok := errno.AsBusinessError(err); ok {
			common.Error(c, bizErr.Code, bizErr.Message)
			return
		}
		common.Error(c, errno.SystemError.Code, errno.SystemError.Message)
		return
	}
	c.JSON(200, response)
}

func (i *ImageController) GetMyRecords(c *gin.Context) {
	loginUser, err := i.userService.GetLoginUser(c)
	if err != nil {
		if bizErr, ok := errno.AsBusinessError(err); ok {
			common.Error(c, bizErr.Code, bizErr.Message)
			return
		}
		common.Error(c, errno.SystemError.Code, errno.SystemError.Message)
		return
	}
	pageNum := int64(1)
	pageSize := int64(10)
	if raw := c.Query("pageNum"); raw != "" {
		if parsed, parseErr := strconv.ParseInt(raw, 10, 64); parseErr == nil && parsed > 0 {
			pageNum = parsed
		}
	}
	if raw := c.Query("pageSize"); raw != "" {
		if parsed, parseErr := strconv.ParseInt(raw, 10, 64); parseErr == nil && parsed > 0 {
			pageSize = parsed
		}
	}
	page, serviceErr := i.imageService.ListUserRecords(loginUser.ID, pageNum, pageSize)
	if serviceErr != nil {
		if bizErr, ok := errno.AsBusinessError(serviceErr); ok {
			common.Error(c, bizErr.Code, bizErr.Message)
			return
		}
		common.Error(c, errno.SystemError.Code, errno.SystemError.Message)
		return
	}
	common.Success(c, page)
}
